runtime/j9vm/javanextvmi.cpp

```diff
@@ -577,13 +577,28 @@ JNIEXPORT void JNICALL
 JVM_VirtualThreadHideFrames(JNIEnv *env, jobject vthread, jboolean hide)
 {
 	J9VMThread *currentThread = (J9VMThread *)env;
+	J9InternalVMFunctions const * const vmFuncs = currentThread->javaVM->internalVMFunctions;
+
+	vmFuncs->internalEnterVMFromJNI(currentThread);
 
+	j9object_t vThreadObj = currentThread->threadObject;
+	Assert_SC_true(IS_JAVA_LANG_VIRTUALTHREAD(currentThread, vThreadObj));
+	/* Do not allow JVMTI operations because J9VMThread->threadObject is modified
+	 * between the first invocation with hide=true and the second invocation with
+	 * hide=false. Otherwise, JVMTI functions will see an unstable
+	 * J9VMThread->threadObject.
+	 */
+	bool hiddenFrames = J9_ARE_ALL_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES);
 	if (hide) {
-		Assert_SC_true(J9_ARE_NO_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES));
+		Assert_SC_true(!hiddenFrames && (vThreadObj == J9_JNI_UNWRAP_REFERENCE(vthread)));
+		enterVThreadTransitionCritical(currentThread, vthread);
 	} else {
-		Assert_SC_true(J9_ARE_ALL_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES));
+		Assert_SC_true(hiddenFrames);
+		exitVThreadTransitionCritical(currentThread, vThreadObj);
 	}
+
 	VM_VMHelpers::virtualThreadHideFrames(currentThread, hide);
```

Comment:
flag toggle should be moved inside the if check to maintain correct order and avoid sync issues in the future
ie:
enterVThreadTransitionCritical
virtualThreadHideFrames(true)
virtualThreadHideFrames(false)
exitVThreadTransitionCritical

Reply:
Fixed `JVM_VirtualThreadHideFrames`. Also, addressed a similar issue in `virtualThreadMountBegin`.